### PR TITLE
Resources S3 storage optimization

### PR DESCRIPTION
**Pull request by michael-topchiev**

### CODE REVIEW

File: app/storage/legacyS3ResourceHandler.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+const deserializer = (urlString) => {
+  const urlObj = new URL(urlString);
+  const fullPath = urlObj.pathname;
+  const parts = _.filter(_.split(fullPath, '/'));
+  const bucketName = parts.shift();
+  const path = `${parts.join('/')}`;
+  return new S3LegacyResourceHandler(decodeURIComponent(path), bucketName, urlString); // endpoint from urlString is not used
+};
+
+module.exports = { constructor, deserializer };
```
Comment (adamkingit) [inline app/storage/legacyS3ResourceHandler.js:111]:
```suggestion
module.exports = { constructor, deserializer };

```

File: app/storage/resourceStorageHandler.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  async deleteData() {
+    return this.targetHandler.deleteData();
+  }
+
+  serialize() {
+    const data = this.targetHandler.serialize();
+    return { metadata: this.metadata, data };
+  }
+};
```
Comment (adamkingit) [inline app/storage/resourceStorageHandler.js:53]:
```suggestion
};

```

File: app/storage/s3ResourceHandler.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+const constructor = (resourceKey, bucketName, location) => {
+  return new S3ResourceHandler(resourceKey, bucketName, location);
+};
+
+const deserializer = (data) => {
+  return new S3ResourceHandler(data.path, data.bucketName, data.location, data.endpoint);
+};
+
+module.exports = { constructor, deserializer };
```
Comment (adamkingit) [inline app/storage/s3ResourceHandler.js:106]:
```suggestion
module.exports = { constructor, deserializer };

```

File: app/storage/s3ResourceHandler.tests.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    const resource2 = { metadata: { type: null }, data: {} };
+    expect(() => storageFactory.deserialize(resource2)).to.throw('Invalid metadata structure');
+
+    const resource3 = { metadata: { type: 'abcd' }, data: {} };
+    expect(() => storageFactory.deserialize(resource3)).to.throw('Resource handler implementation for type abcd is not defined');
+
+  });
+
+});
```
Comment (adamkingit) [inline app/storage/s3ResourceHandler.tests.js:104]:
```suggestion
});

```

File: app/storage/s3ResourceHandlerTest.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  const longString = 'x'.repeat(100*1024*1024);
+  const handler2 = storageFactory.newResourceHandler(path, bucketName, location);
+  const promise = handler2.setData(longString);
+  console.log('Waiting on promise...' + new Date());
+  await promise;
+  handler2.deleteData();
+  console.log('All done!');
+}
+
+test().catch(console.log);
```
Comment (adamkingit) [inline app/storage/s3ResourceHandlerTest.js:64]:
```suggestion
test().catch(console.log);

```

File: app/storage/storageFactory.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  logInfo(msg) {
+    logger.info(msg);
+  }
+
+  isLink(s) {
+    return _.isString(s) && /^(http|https):\/\/?/.test(s);
+  }
+}
+
+module.exports = new StorageFactory();
```
Comment (adamkingit) [inline app/storage/storageFactory.js:86]:
```suggestion
module.exports = new StorageFactory();

```

File: app/storage/embeddedResourceHandlerTest.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  const handler2 = storageFactory.newResourceHandler(path, bucketName, location);
+  const promise = handler2.setData(longString);
+  console.log('Waiting on promise...' + new Date());
+  await promise;
+  handler2.deleteData();
+  console.log('All done!');
+
+}
+
+test().catch(console.log);
```
Comment (adamkingit) [inline app/storage/embeddedResourceHandlerTest.js:54]:
```suggestion
test().catch(console.log);

```

File: app/storage/embeddedResourceHandler.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+const constructor = () => {
+  return new EmbeddedResourceHandler();
+};
+
+const deserializer = (data) => {
+  return new EmbeddedResourceHandler(data);
+};
+
+module.exports = { constructor, deserializer };
```
Comment (adamkingit) [inline app/storage/embeddedResourceHandler.js:69]:
```suggestion
module.exports = { constructor, deserializer };

```

File: app/storage/cipher.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    if (_.isString(iv)) {
+        iv = Buffer.from(iv, 'base64');
+    }
+    const key = Buffer.concat([Buffer.from(encryptionKey)], 32);
+    const decipher = crypto.createDecipheriv(encryptionAlgorithm, key, iv);
+    return { decryptedStream: encryptedStream.pipe(decipher) };
+}
+*/
+
+module.exports = { encrypt, decryptBuffer };  // encryptStream, decryptStream
```
Comment (adamkingit) [inline app/storage/cipher.js:67]:
```suggestion
module.exports = { encrypt, decryptBuffer };  // encryptStream, decryptStream

```

File: app/storage/cipher.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+'use strict';
+
+const crypto = require('crypto');
+const _ = require('lodash');
+
+const encryptionAlgorithm = 'aes-256-cbc';
+
+const encrypt = (stringOrBuffer, encryptionKey) => {
```
Comment (adamkingit) [inline app/storage/cipher.js:24]: Is this new encryption, or a refactoring of what was already there?

**Review by rmgraham — APPROVED** [review]

**Review by carrolp — DISMISSED** [review]

**Comment by carrolp** [comment]
Added approval: combined rmgraham's earlier approval with @michael-topchiev 's approval of my added commit that dismissed rmgraham's.... == approved again.

**Review by dalehille — DISMISSED** [review]

**Review by dalehille — APPROVED** [review]

**Pull request merged**